docs/design/datacontracts/data_descriptor.md

```diff
@@ -0,0 +1,264 @@
+
+Rationale: If a type appears in multiple physical descriptors, the later appearances may add more
+fields or change the offsets or definite/indefinite sizes of prior definitions.  If a value appears
+multiple times, later definitions take precedence.
+
+## Physical JSON descriptor
+
+### Version
+
+This is version 0 of the physical descriptor
+
+### Summary
+
+A data descriptor may be stored in the "JSON with comments" format.
+
+The toplevel dictionary will contain:
+
+* `"version": 0`
+* `"types": TYPE_ARRAY` see below
+* `"globals": VALUE_ARRAY` see below
+
+### Types
+
+The types will be in an array, with each type described by a dictionary containing keys:
+
+* `"name": "type name"` the name of each type
+* optional `"size": int | "indeterminate"` if omitted the size is indeterminate
+* optional `"fields": FIELD_ARRAY` if omitted same as a field array of length zero
+
+Each `FIELD_ARRAY` is an array of dictionaries each containing keys:
+
+* `"name": "field name"` the name of each field
+* `"type": "type name"` the name of a primitive type or another type defined in the logical descriptor
+* optional `"offset": int | "unknown"` the offset of the field or "unknown". If omitted, same as "unknown".
+
+Note that the logical descriptor does not contain "unknown" offsets: it is expected that the binary
+blob will augment the baseline with a known offset for all fields in the baseline.
+
+Rationale: "unknown" offsets may be used to document in the physical JSON descriptor that the binary
+blob descriptor is expected to provide the offset of the field.
+
+### Global values
+
+The global values will be in an array, with each value described by a dictionary containing keys:
+
+* `"name": "global value name"` the name of the global value
+* `"type": "type name"` the type of the global value
+* optional `"value": VALUE | { "indirect": int } | "unknown"` the value of the global value, or an offset in an auxiliary array containing the value or "unknown".
+
+Note that the logical descriptor does not contain "unknown" values: it is expected that the binary
+blob will augment the baseline with a known offset for all fields in the baseline.
+
+The `VALUE` may be a JSON numeric constant integer or a string containing a signed or unsigned
+decimal or hex (with prefix `0x` or `0X`) integer constant.  The constant must be within the range
+of the type of the global value.
+
+For pointer and nuint globals, the value may be assumed to fit in a 64-bit unsigned integer.  For
+nint globals, the value may be assumed to fit in a 64-bit signed integer.
+
+If the value is given as `{"indirect": int}` then the value is stored in an auxiliary array that is
+part of the data contrat descriptor.  Only in-memory data descriptors may have indirect values; baseline data descriptors may not have indirect values.
+
+Rationale: This allows tooling to generate the in-memory data descriptor as a single constant
```

Comment:
I was under the impression that the JSON format was for documentation purposes only and it would never appear in memory of the target process. What is the scenario where we expect the JSON format to be embedded in memory?

Reply:
> I was under the impression that the JSON format was for documentation purposes only

The well-known baseline JSON data will be embedded into the CDAC reader tool. It's not just for documentation.

> What is the scenario where we expect the JSON format to be embedded in memory?

If the in-proc data descriptor is going to be a data-segment constant that is being generated by `cdac-build-tool`, there didn't seem to be any reason why we should be a binary blob.  JSON is easy to read with a variety of off-the-shelf parsers.

Reply:
Also:
- The size of the compact JSON format is not that much different from the size of the originally proposed binary format, and there are standardized ways to make JSON smaller if we need to (compression, binary JSON, ...)
- Designing our own binary format for structured data is a rathole

Reply:
I just misunderstood the text. I thought it was saying the JSON would be inside the target process being debugged. If this JSON is included in the reader then I have no concern with that at all. I consider it an implementation choice of any diagnostic tool, including ours, how it wants to utilize the documentation. Sorry for confusion!

Reply:
> I thought it was saying the JSON would be inside the target process being debugged. 

The text is saying that the (compact) JSON will be inside the runtime binary in the target process.

Reply:
Sorry it looks like I had it right the first time. I see there was an edit earlier today that eliminated the binary format entirely and just use the JSON as the in-memory format.

This feels like it is getting progressively further away from the approach I would have favored, but I think you can make it work and it feels like my opinion on this is in the minority. My impression is that defining flat in-memory structs with no baseline at all would be easier to generate (no build tooling needed or baseline to manage), easier to consume (no JSON parser needed), and fits in our size goals. As an example the GC already does [this](https://github.com/dotnet/runtime/blob/main/src/coreclr/gc/gcinterface.dac.h#L231) as have past successful prototypes. I know I suggested having a baseline at an earlier point, but pretty quickly it felt like the amount of complexity it added to the design made it feel not worth the potential size savings.

In any case if your vision is different and you think JSON, baselines, and custom build tools are the best approach go for it. I see no reason it won't work even if it isn't the approach I would have picked.

Reply:
> it felt like the amount of complexity it added to the design made it feel not worth the potential size savings.

@lambdageek and I chatted about this too. I'm hoping we can focus more on actually understanding what the entire description will be and then after we know for sure we can apply a baseline logic or just accept that compisition wasn't needed because the size isn't as bad as assumed.

Reply:
Part of my thinking that we wouldn't need a baseline is that I'm expecting a simple ordinal based binary encoding to be between 10-40% of the size of JSON encoding for the same set of data. For example you might see that a JSON encoding of the entire runtime descriptor with no baseline is 20KB in size which we decide is too big. Then we either save size by sticking with JSON and using a baseline, or we save size by using some simple structs with binary data that is substantially more efficient.